eth/sync.go

```diff
@@ -182,19 +179,46 @@ func (pm *ProtocolManager) synchronise(peer *peer) {
 	// Otherwise try to sync with the downloader
 	mode := downloader.FullSync
 	if atomic.LoadUint32(&pm.fastSync) == 1 {
+		// Fast sync was explicitly requested, and explicitly granted
 		mode = downloader.FastSync
+	} else if currentBlock.NumberU64() == 0 && pm.blockchain.CurrentFastBlock().NumberU64() > 0 {
+		// The database seems empty as the current block is the genesis. Yet the fast
+		// block is ahead, so fast sync was enabled for this node at a certain point.
+		// The only scenario where this can happen is if the user manually (or via a
+		// bad block) rolled back a fast sync node below the sync point. In this case
+		// however it's safe to reenable fast sync.
+		atomic.StoreUint32(&pm.fastSync, 1)
+		mode = downloader.FastSync
```

Comment:
Here's a new behavior; if you run `geth --fast` and _don't progress thru downloading full blocks_ before restarting, then if you run just plain `geth`, fast sync will resume even if you don't pass the flag. This got me. ⚡️ 👀

Reply:
Strange behaviour. So in such case to reastart geth with full sync I need to delete db?

Reply:
As-is, yes. Or just continue syncing until you reach a pivot point, then it will still automatically switch to full sync and then, like usual, always full sync from there on out.

I agree that it's kind of strange, but maybe that's because we've become so used to being very careful about the `--fast` flag. 

And it is kind of a nice thing to have; for example last week I accidentally made a friend have to restart syncing from 0 because my command instructions for him left the `--fast` off and he started accidentally full syncing around block 3mm.

Overall, I guess there would be fewer cases when someone wants to start full syncing from a random block in the middle of the chain, compared to when someone starts with `--fast` and then forgets to use the flag on a restart later.

We could consider adding a `--full` flag later, maybe? 😜